### PR TITLE
🌱 TK-5288: CI changes to build and push preflight image

### PR DESCRIPTION
**Pull request by himanshu-trilio**
- same as title

### CODE REVIEW

File: .github/workflows/plugin-packages.yml [hunk 1]
```diff
@@ -190,6 +196,51 @@ jobs:
...
+          export_default_credentials: true
+
+      # Configure docker to use the gcloud command-line tool as a credential helper
+      - name: Set up docker auth
+        run: |
+          gcloud auth configure-docker
+
+      # This step is run when push is not for a tag
+      - name: Sets image tag env to private registry for internal build
+        if: (needs.pre-build.outputs.testPreflight == 'true') && !startsWith(github.ref, 'refs/tags/')
```
Comment (msurudwad) [inline .github/workflows/plugin-packages.yml:231]:
```suggestion
        if: (needs.pre-build.outputs.preflight == 'true') && !startsWith(github.ref, 'refs/tags/')
```

File: .github/workflows/plugin-packages.yml [hunk 2]
```diff
@@ -190,6 +196,51 @@ jobs:
...
+        run: echo "DOCKER_REGISTRY=$GCR_PRIVATE_DOCKER_REGISTRY" >> $GITHUB_ENV
+
+      # This step is run when push is for a tag
+      - name: Sets image tag env to public registry for release
+        if: (needs.pre-build.outputs.testPreflight == 'true') && startsWith(github.ref, 'refs/tags/')
```
Comment (msurudwad) [inline .github/workflows/plugin-packages.yml:236]:
```suggestion
        if: (needs.pre-build.outputs.preflight == 'true') && startsWith(github.ref, 'refs/tags/')
```

File: .github/workflows/plugin-packages.yml (hunk 1 above)
Comment (msurudwad) [inline .github/workflows/plugin-packages.yml:231]: does `!startsWith` syntax work? For earlier versions, this exact syntax was not supported.

File: .github/workflows/plugin-packages.yml [hunk 3]
```diff
@@ -190,6 +196,51 @@ jobs:
...
+        run: echo "DOCKER_REGISTRY=$GCR_PUBLIC_DOCKER_REGISTRY" >> $GITHUB_ENV
+
+      # Build and push the Docker image
+      - name: Build and push image
```
Comment (msurudwad) [inline .github/workflows/plugin-packages.yml:240]: try to use pre-build github actions for docker build & push. Refer [doc](https://docs.docker.com/ci-cd/github-actions/#set-up-the-github-actions-workflow).. also check the https://docs.docker.com/ci-cd/github-actions/#optimizing-the-workflow

File: .github/workflows/plugin-packages.yml [hunk 4]
```diff
@@ -190,6 +196,51 @@ jobs:
...
+        run: |
+          docker build -t ${{env.DOCKER_REGISTRY}}/${{env.GCR_PROJECT}}/${{env.IMAGE_NAME}}:latest -f docker-images/preflight/Dockerfile .
```
Comment (msurudwad) [inline .github/workflows/plugin-packages.yml:242]: `latest` tag cannot be used for all pushes - it'll be overridden.
for normal branches push - image tag should that branch's commit SHA
for master branch push - image tag should that branch's commit SHA & `latest`
for git tags push - image tag should that git tag and `latest`

**Review by msurudwad — CHANGES_REQUESTED** [review]
Update pre-build job's `Detect File Changes` step so that CI jobs will run on change to `docker-images/preflight/Dockerfile` file

File: .github/workflows/plugin-packages.yml (hunk 1 above)
Comment (himanshu-trilio) [reply .github/workflows/plugin-packages.yml]: yes. tried this on the current workflow as well.

Comment (himanshu-trilio) [reply .github/workflows/plugin-packages.yml]: changed

**Comment by himanshu-trilio** [comment]
> Update pre-build job's `Detect File Changes` step so that CI jobs will run on change to `docker-images/preflight/Dockerfile` file

updated

File: .github/workflows/plugin-packages.yml (hunk 3 above)
Comment (himanshu-trilio) [reply .github/workflows/plugin-packages.yml]: updated yaml to use github-actions directives

File: .github/workflows/plugin-packages.yml (hunk 4 above)
Comment (himanshu-trilio) [reply .github/workflows/plugin-packages.yml]: used this: https://github.com/marketplace/actions/docker-metadata-action for defining the image tags and also updated yaml according to this configuration. not able to test on tag or branch push but works fine or PR push

File: .github/workflows/plugin-packages.yml [hunk 5]
```diff
@@ -185,11 +192,73 @@ jobs:
...
+
+      - name: Docker meta
+        id: meta
+        uses: docker/metadata-action@v3
+        with:
+          images: ${{env.DOCKER_REGISTRY}}/${{env.GCR_PROJECT}}/${{env.IMAGE_NAME}}
+          tags: |
+            type=sha,prefix=,format=short,event=pr
+            type=ref,event=branch
+            type=sha,prefix=,format=short,event=branch
```
Comment (msurudwad) [inline .github/workflows/plugin-packages.yml:250]: use `format=long`

File: .github/workflows/plugin-packages.yml [hunk 6]
```diff
@@ -185,11 +192,73 @@ jobs:
...
+        run: echo "DOCKER_REGISTRY=$GCR_PUBLIC_DOCKER_REGISTRY" >> $GITHUB_ENV
+
+      - name: Docker meta
+        id: meta
+        uses: docker/metadata-action@v3
+        with:
+          images: ${{env.DOCKER_REGISTRY}}/${{env.GCR_PROJECT}}/${{env.IMAGE_NAME}}
+          tags: |
+            type=sha,prefix=,format=short,event=pr
+            type=ref,event=branch
```
Comment (msurudwad) [inline .github/workflows/plugin-packages.yml:249]: check if this line is required?

File: .github/workflows/plugin-packages.yml [hunk 7]
```diff
@@ -185,11 +192,73 @@ jobs:
...
+          tools/preflight/junit-preflight-unit-test.xml
 
     - name: Test Preflight Plugin Locally
       run: make test-preflight-plugin-locally
 
+  # build_push_preflight_image job builds preflight docker image, and pushes it to public docker registry,
+  # if changes for tags('/ref/tags') are pushed or otherwise pushes to private docker registry
+  build_push_preflight_image:
+    runs-on: ubuntu-latest
+    needs: [pre-build, build_test_preflight]
```
Comment (msurudwad) [inline .github/workflows/plugin-packages.yml:206]: remove `build_test_preflight`

**Review by msurudwad — CHANGES_REQUESTED** [review]

**Review by msurudwad — APPROVED** [review]

File: .github/workflows/plugin-packages.yml [hunk 8]
```diff
@@ -199,6 +199,77 @@ jobs:
...
+      IMAGE_NAME: "preflight"
+    steps:
+
+      # check out the repository
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      # provide emulated environments for building on architecture other than the host arch
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@master
```
Comment (umeshkumhar) [inline .github/workflows/plugin-packages.yml:223]: remove master and keep latest version

File: .github/workflows/plugin-packages.yml [hunk 9]
```diff
@@ -199,6 +199,77 @@ jobs:
...
+        with:
+          platforms: linux/amd64,linux/arm64,linux/ppc64le
+
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@master
```
Comment (umeshkumhar) [inline .github/workflows/plugin-packages.yml:229]: same here

**Review by umeshkumhar — APPROVED** [review]

File: .github/workflows/plugin-packages.yml (hunk 8 above)
Comment (himanshu-trilio) [reply .github/workflows/plugin-packages.yml]: done

File: .github/workflows/plugin-packages.yml (hunk 9 above)
Comment (himanshu-trilio) [reply .github/workflows/plugin-packages.yml]: done

**Pull request merged**